File: src/main/java/xf/xflp/opt/FastMinContainerSolver.java
package xf.xflp.opt;

import xf.xflp.base.XFLPModel;
import xf.xflp.base.item.Item;
import xf.xflp.exception.XFLPException;
import xf.xflp.opt.construction.multitype.NContainerNTypeAddPacker;
import xf.xflp.opt.construction.onetype.NContainerOneTypeAddPacker;
import xf.xflp.report.LoadType;

/**
 * Copyright (c) 2012-2022 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @author hschneid
 *
 * This solver tries to create very fastly a reasonable solution.
 *
 * It uses only construction heuristics.
 *
 * Goal: Place all items into a some instances of container types. The number of containers
 *       shall be minimal.
 */
public class FastMinContainerSolver extends XFLPBase {

    private final NContainerOneTypeAddPacker oneTypeAddPacker = new NContainerOneTypeAddPacker();
    private final NContainerNTypeAddPacker nTypeAddPacker = new NContainerNTypeAddPacker();

    @Override
    public void execute(XFLPModel model) throws XFLPException {
        if(isOnlyAddingItems(model)) {
            if(model.getContainerTypes().length > 1) {
                nTypeAddPacker.execute(model);
            } else {
                oneTypeAddPacker.execute(model);
            }
        } else {
            if(model.getContainerTypes().length > 1) {
                throw new UnsupportedOperationException("Currently add/removing and multiple container types is not supported");
            } else {
                throw new UnsupportedOperationException("Currently add/removing and single container types is not supported");
            }
        }
    }

    private boolean isOnlyAddingItems(XFLPModel model) {
        for (Item item : model.getItems()) {
            if(item.loadingType == LoadType.UNLOAD) {
                return false;
            }
        }

        return true;
    }
}

File: src/xflp/opt/fast_min_container_solver.py
# Copyright (c) 2012-2025 Holger Schneider
# All rights reserved.
#
# This source code is licensed under the MIT License (MIT) found in the
# LICENSE file in the root directory of this source tree.

"""
This solver tries to create very fast a reasonable solution.

It uses only construction heuristics.

Goal: Place all items into some instances of container types. The number of containers
      shall be minimal.

Author: hschneid
"""

from typing import TYPE_CHECKING

from xflp.opt.xflp_base import XFLPBase
from xflp.opt.construction.onetype.n_container_one_type_add_packer import NContainerOneTypeAddPacker
from xflp.opt.construction.multitype.n_container_n_type_add_packer import NContainerNTypeAddPacker
from xflp.report.load_type import LoadType

if TYPE_CHECKING:
    from xflp.base.xflp_model import XFLPModel


class FastMinContainerSolver(XFLPBase):
    """Fast minimum container solver using construction heuristics."""

    def __init__(self):
        self.one_type_add_packer = NContainerOneTypeAddPacker()
        self.n_type_add_packer = NContainerNTypeAddPacker()

    def execute(self, model: 'XFLPModel') -> None:
        """
        Execute the fast minimum container solver.

        Args:
            model: The XFLP model to optimize

        Raises:
            XFLPException: If optimization fails
        """
        if self._is_only_adding_items(model):
            if len(model.get_container_types()) > 1:
                self.n_type_add_packer.execute(model)
            else:
                self.one_type_add_packer.execute(model)
        else:
            if len(model.get_container_types()) > 1:
                raise NotImplementedError("Currently add/removing and multiple container types is not supported")
            else:
                raise NotImplementedError("Currently add/removing and single container types is not supported")

    def _is_only_adding_items(self, model: 'XFLPModel') -> bool:
        """Check if all items are only being added (not unloaded)."""
        for item in model.get_items():
            if item.loading_type == LoadType.UNLOAD:
                return False
        return True
